File: sdk/python/forkd/controller.py
"""HTTP client for the forkd-controller daemon's REST API.

The `Controller` class wraps `/v1/snapshots` and `/v1/sandboxes` endpoints
(docs/API.md) so Python agent code can manage snapshots, fork sandboxes,
branch running sandboxes, and tear things down without shelling out to
`forkd`.

Talking to the controller is orthogonal to the in-guest `Sandbox` agent
class (`forkd.Sandbox`): Controller manages VM lifecycle from the host
side; Sandbox talks to the in-guest agent on TCP for exec/eval inside
one specific child VM. Most agent runtimes use both — Controller to
spawn/branch/kill, Sandbox to drive code execution.

Example
-------

>>> from forkd import Controller, Sandbox
>>> c = Controller()  # default http://127.0.0.1:8889, no token
>>> [s["tag"] for s in c.list_snapshots()]
['pyagent']
>>> children = c.spawn_sandboxes("pyagent", n=1, per_child_netns=True)
>>> sb_id = children[0]["id"]
>>> # ... drive the sandbox via Sandbox(target=children[0]['guest_addr'])
>>> branch = c.branch_sandbox(sb_id, tag="checkpoint-1")
>>> branch["tag"]
'checkpoint-1'
>>> branch["branched_from"]
'sb-...'
>>> grandchildren = c.spawn_sandboxes(branch["tag"], n=5)
>>> c.kill_sandbox(sb_id)
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from typing import Any, Literal, Optional

BranchMode = Literal["full", "diff", "live"]
"""Canonical BRANCH mode selector (Phase 7.1+).

- ``"full"`` — copy entire guest RAM under pause (default for v0.x).
- ``"diff"`` — Firecracker Diff snapshot (v0.3+). Sub-second pause for
  idle sources; replaces the legacy ``diff=True`` boolean.
- ``"live"`` — UFFD_WP-based live BRANCH (v0.4+). Source pause drops to
  sub-50 ms; memory streams from the running parent. Requires the
  source to have been spawned with ``live_fork=True``.
"""


class ControllerError(RuntimeError):
    """Raised on non-2xx responses from the daemon.

    Carries the HTTP status code and the daemon's parsed error body
    (when it returned JSON). Inspect ``status`` and ``body`` to
    distinguish 404 (sandbox/snapshot missing) from 409 (tag
    collision) from 500 (internal).
    """

    def __init__(self, status: int, body: Any, url: str) -> None:
        self.status = status
        self.body = body
        self.url = url
        snippet = body if isinstance(body, str) else json.dumps(body)
        super().__init__(f"controller {url}: HTTP {status}: {snippet}")


class Controller:
    """Client for the forkd-controller daemon's REST API.

    Parameters
    ----------
    base_url:
        Daemon base URL. Defaults to ``$FORKD_URL`` then
        ``http://127.0.0.1:8889``.
    token:
        Bearer token. Defaults to ``$FORKD_TOKEN``. Required only when
        the daemon was started with ``--token-file``.
    timeout:
        Per-request timeout in seconds. Branching can take 0.5-8 s on
        a large parent VM; default is generous.
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        token: Optional[str] = None,
        timeout: float = 60.0,
    ) -> None:
        self.base_url = (
            base_url
            or os.environ.get("FORKD_URL")
            or "http://127.0.0.1:8889"
        ).rstrip("/")
        self.token = token if token is not None else os.environ.get("FORKD_TOKEN")
        self.timeout = timeout

    # --- snapshots ------------------------------------------------

    def list_snapshots(self) -> list[dict]:
        """``GET /v1/snapshots`` — every snapshot known to the daemon."""
        return self._request("GET", "/v1/snapshots")

    def delete_snapshot(self, tag: str) -> None:
        """``DELETE /v1/snapshots/:tag`` — drop both registry and disk files."""
        self._request("DELETE", f"/v1/snapshots/{tag}")

    # --- sandboxes ------------------------------------------------

    def spawn_sandboxes(
        self,
        snapshot_tag: str,
        n: int = 1,
        per_child_netns: bool = False,
        memory_limit_mib: Optional[int] = None,
        prewarm: bool = False,
        live_fork: bool = False,
        hugepages: bool = False,
    ) -> list[dict]:
        """``POST /v1/sandboxes`` — fork N children from a snapshot tag.

        Parameters
        ----------
        prewarm:
            When true, each child performs a throwaway snapshot to
            scratch storage immediately after restore to fault-in all
            guest pages. Trades ~170 ms / 512 MiB of extra spawn time
            for steady-state BRANCH latency on the first user-visible
            BRANCH (avoids the 2-9× cold-cache penalty documented in
            ``bench/pause-window/RESULTS-v0.2.md``).
        live_fork:
            v0.4+. Boot the sandbox with a memfd-backed RAM region so
            later BRANCHes from it can use ``mode="live"`` (UFFD_WP).
            Requires kernel 5.7+ and the vendored Firecracker fork —
            see ``docs/VENDORED-FIRECRACKER.md``. No effect at spawn
            time beyond the backend swap; cost shows up on the first
            live BRANCH.
        hugepages:
            v0.4+. Back the memfd with 2 MiB hugepages
            (``MFD_HUGETLB | MFD_HUGE_2MB``). Only meaningful with
            ``live_fork=True``. Reduces TLB pressure during spawn-many
            and live BRANCH bulk-copy. Requires non-zero
            ``HugePages_Free`` in ``/proc/meminfo`` — ``forkd doctor``
            checks availability. Falls back to normal 4 KiB pages with
            a warning if the pool is exhausted.

        Returns the list of SandboxInfo dicts (id, snapshot_tag, netns,
        guest_addr, created_at_unix, pid, memory_limit_mib).
        """
        body: dict[str, Any] = {
            "snapshot_tag": snapshot_tag,
            "n": n,
            "per_child_netns": per_child_netns,
        }
        if memory_limit_mib is not None:
            body["memory_limit_mib"] = memory_limit_mib
        if prewarm:
            body["prewarm"] = True
        if live_fork:
            body["live_fork"] = True
        if hugepages:
            body["hugepages"] = True
        return self._request("POST", "/v1/sandboxes", body)

    def list_sandboxes(self) -> list[dict]:
        """``GET /v1/sandboxes`` — every live sandbox the daemon tracks."""
        return self._request("GET", "/v1/sandboxes")

    def get_sandbox(self, sandbox_id: str) -> dict:
        """``GET /v1/sandboxes/:id`` — one sandbox's metadata."""
        return self._request("GET", f"/v1/sandboxes/{sandbox_id}")

    def kill_sandbox(self, sandbox_id: str) -> None:
        """``DELETE /v1/sandboxes/:id`` — terminate one sandbox."""
        self._request("DELETE", f"/v1/sandboxes/{sandbox_id}")

    def branch_sandbox(
        self,
        sandbox_id: str,
        tag: Optional[str] = None,
        diff: bool = False,
        measure_diff: bool = False,
        mode: Optional[BranchMode] = None,
        wait: bool = True,
    ) -> dict:
        """``POST /v1/sandboxes/:id/branch`` — pause + snapshot + resume.

        Parameters
        ----------
        mode:
            v0.4+ canonical selector. ``"full"``, ``"diff"``, or
            ``"live"``. When set, takes precedence over the legacy
            ``diff`` boolean — and passing both raises
            :class:`ControllerError` (HTTP 400). Prefer this over
            ``diff=`` in new code. See :data:`BranchMode`.
        diff:
            **Legacy.** Equivalent to ``mode="diff"``; kept so this SDK
            can drive v0.3.x daemons that don't understand ``mode``.
            Mutually exclusive with ``mode`` (server-side).
        measure_diff:
            v0.3+: measurement-only hook. Take a Diff snapshot inside
            the existing Full pause to report what diff would have
            cost, without changing semantics. Mutually exclusive with
            ``diff`` (daemon returns 400 if both are true).
        wait:
            v0.4+, only meaningful with ``mode="live"``. Default
            ``True`` blocks until the background memory copy finishes
            and the returned snapshot is ``status="ready"``. Set to
            ``False`` to return as soon as the source resumes (~10 ms);
            the snapshot reaches ``status="ready"`` later — poll
            :meth:`list_snapshots` to detect completion.

        The source sandbox is paused for the duration of the snapshot
        write — typically 0.5-8 s for Full, ~200 ms for Diff, sub-50 ms
        for Live — then resumed. The returned snapshot is independent
        of the source's lifecycle.

        Returns a SnapshotInfo dict; pass its ``tag`` to
        ``spawn_sandboxes`` to fork grandchildren from the branch.
        """
        body: dict[str, Any] = {}
        if tag is not None:
            body["tag"] = tag
        # Prefer canonical `mode` when set; fall back to legacy `diff`
        # so older daemons keep working unchanged.
        if mode is not None:
            body["mode"] = mode
        elif diff:
            body["diff"] = True
        if measure_diff:
            body["measure_diff"] = True
        # `wait=True` is the daemon default; only send when the caller
        # opted into fire-and-forget so the body stays minimal against
        # daemons that don't recognize the field.
        if not wait:
            body["wait"] = False
        return self._request("POST", f"/v1/sandboxes/{sandbox_id}/branch", body)

    def exec_command(
        self,
        sandbox_id: str,
        args: list[str],
        timeout_secs: int = 30,
    ) -> dict:
        """``POST /v1/sandboxes/:id/exec`` — run a subprocess in the sandbox.

        Returns ``{stdout, stderr, exit_code}``.
        """
        return self._request(
            "POST",
            f"/v1/sandboxes/{sandbox_id}/exec",
            {"args": args, "timeout_secs": timeout_secs},
        )

    def eval_code(self, sandbox_id: str, code: str) -> dict:
        """``POST /v1/sandboxes/:id/eval`` — eval against warmed PID-1.

        Returns ``{result, error, exit_code}``.
        """
        return self._request(
            "POST",
            f"/v1/sandboxes/{sandbox_id}/eval",
            {"code": code},
        )

    def ping_sandbox(self, sandbox_id: str) -> dict:
        """``POST /v1/sandboxes/:id/ping`` — round-trip to the guest agent."""
        return self._request("POST", f"/v1/sandboxes/{sandbox_id}/ping")

    # --- internals ------------------------------------------------

    def _request(self, method: str, path: str, body: Optional[dict] = None) -> Any:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode() if body is not None else None
        headers = {"Content-Type": "application/json"} if body is not None else {}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        req = urllib.request.Request(url, data=data, method=method, headers=headers)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
                if not raw:
                    return None
                return json.loads(raw)
        except urllib.error.HTTPError as e:
            raw = e.read()
            parsed: Any
            try:
                parsed = json.loads(raw) if raw else {}
            except json.JSONDecodeError:
                parsed = raw.decode(errors="replace")
            raise ControllerError(e.code, parsed, url) from e

File: sdk/typescript/src/controller.ts
import type {
  BranchOptions,
  EvalResult,
  ExecOptions,
  ExecResult,
  PingResult,
  SandboxInfo,
  SnapshotInfo,
  SpawnOptions,
} from "./types.js";

/**
 * Raised on non-2xx responses from the daemon.
 *
 * Inspect `status` and `body` to distinguish 404 (sandbox/snapshot
 * missing) from 409 (tag collision) from 500 (internal).
 */
export class ControllerError extends Error {
  readonly status: number;
  readonly body: unknown;
  readonly url: string;

  constructor(status: number, body: unknown, url: string) {
    const snippet =
      typeof body === "string" ? body : JSON.stringify(body);
    super(`controller ${url}: HTTP ${status}: ${snippet}`);
    this.name = "ControllerError";
    this.status = status;
    this.body = body;
    this.url = url;
  }
}

export interface ControllerOptions {
  /**
   * Daemon base URL. Defaults to env `FORKD_URL` then
   * `http://127.0.0.1:8889`.
   */
  baseUrl?: string;
  /**
   * Bearer token. Defaults to env `FORKD_TOKEN`. Required when the
   * daemon was started with `--token-file`.
   */
  token?: string;
  /**
   * Per-request timeout in milliseconds. Default 60_000.
   * Branching a large source can take several seconds before v0.3's
   * diff mode kicks in.
   */
  timeoutMs?: number;
  /**
   * Custom fetch implementation. Defaults to global `fetch`. Provide
   * `undici`'s fetch in Node test scenarios that want to record
   * traffic, or polyfill for older Node.
   */
  fetch?: typeof fetch;
}

/**
 * Client for the forkd-controller daemon's REST API.
 *
 * @example
 * ```ts
 * import { Controller } from '@deeplethe/forkd';
 * const ctrl = new Controller({
 *   baseUrl: 'http://127.0.0.1:8889',
 *   token: process.env.FORKD_TOKEN,
 * });
 * const snapshots = await ctrl.listSnapshots();
 * const [sb] = await ctrl.spawnSandboxes({ snapshotTag: 'python-3-12-slim' });
 * const result = await ctrl.execCommand(sb.id, ['python3', '-c', 'print(2+2)']);
 * const branch = await ctrl.branchSandbox(sb.id, { diff: true });
 * await ctrl.killSandbox(sb.id);
 * ```
 */
export class Controller {
  readonly baseUrl: string;
  readonly token: string | undefined;
  readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: ControllerOptions = {}) {
    const envUrl =
      typeof process !== "undefined" ? process.env.FORKD_URL : undefined;
    const envToken =
      typeof process !== "undefined" ? process.env.FORKD_TOKEN : undefined;
    this.baseUrl = (opts.baseUrl ?? envUrl ?? "http://127.0.0.1:8889").replace(
      /\/+$/,
      "",
    );
    this.token = opts.token ?? envToken ?? undefined;
    this.timeoutMs = opts.timeoutMs ?? 60_000;
    this.fetchImpl = opts.fetch ?? globalThis.fetch;
    if (typeof this.fetchImpl !== "function") {
      throw new Error(
        "Controller: no fetch implementation. Node 18+ ships fetch globally; otherwise pass `fetch` in options.",
      );
    }
  }

  // --- snapshots ----------------------------------------------------

  async listSnapshots(): Promise<SnapshotInfo[]> {
    return this.request<SnapshotInfo[]>("GET", "/v1/snapshots");
  }

  async deleteSnapshot(tag: string): Promise<void> {
    await this.request<null>("DELETE", `/v1/snapshots/${encodeURIComponent(tag)}`);
  }

  // --- sandboxes ----------------------------------------------------

  /**
   * Fork N children from a registered snapshot.
   *
   * @param options.snapshotTag    snake_case in the wire format,
   *                               camelCase here.
   * @param options.prewarm        v0.2.5+. Relocates the cold-cache
   *                               penalty from the first BRANCH to
   *                               sandbox-creation time.
   * @param options.liveFork       v0.4+. Boot with memfd-backed RAM so
   *                               later BRANCHes from this sandbox can
   *                               use `mode: "live"`. Requires kernel
   *                               5.7+ and the vendored Firecracker
   *                               fork.
   */
  async spawnSandboxes(options: {
    snapshotTag: string;
    n?: number;
    perChildNetns?: boolean;
    memoryLimitMib?: number;
    prewarm?: boolean;
    liveFork?: boolean;
    /** v0.4+: back the memfd with 2 MiB hugepages. Only meaningful with `liveFork: true`. */
    hugepages?: boolean;
  }): Promise<SandboxInfo[]> {
    const body: SpawnOptions = {
      snapshot_tag: options.snapshotTag,
      n: options.n ?? 1,
      per_child_netns: options.perChildNetns ?? false,
    };
    if (options.memoryLimitMib !== undefined) {
      body.memory_limit_mib = options.memoryLimitMib;
    }
    if (options.prewarm !== undefined) {
      body.prewarm = options.prewarm;
    }
    if (options.liveFork !== undefined) {
      body.live_fork = options.liveFork;
    }
    if (options.hugepages !== undefined) {
      body.hugepages = options.hugepages;
    }
    return this.request<SandboxInfo[]>("POST", "/v1/sandboxes", body);
  }

  async listSandboxes(): Promise<SandboxInfo[]> {
    return this.request<SandboxInfo[]>("GET", "/v1/sandboxes");
  }

  async getSandbox(sandboxId: string): Promise<SandboxInfo> {
    return this.request<SandboxInfo>(
      "GET",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}`,
    );
  }

  async killSandbox(sandboxId: string): Promise<void> {
    await this.request<null>(
      "DELETE",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}`,
    );
  }

  /**
   * Branch a running sandbox into a new snapshot.
   *
   * Pauses the source briefly, snapshots, resumes. Pause window
   * depends on `options.mode`:
   *
   * - `"full"` (default): 0.5-8 s, whole guest RAM written.
   * - `"diff"` (v0.3+): ~200 ms idle source, 6-15× speedup on typical
   *   agent workloads, 143× ceiling on 4 GiB SSD.
   * - `"live"` (v0.4+): sub-50 ms; memory streams from the running
   *   parent via UFFD_WP. Requires source booted with
   *   `liveFork: true`. Combine with `wait: false` to return after
   *   the source resumes (~10 ms) without waiting on the background
   *   copy.
   *
   * The legacy `options.diff` boolean still works for v0.3.x daemon
   * compat but is mutually exclusive with `options.mode` server-side.
   *
   * Returns a {@link SnapshotInfo}; pass its `tag` back into
   * {@link spawnSandboxes} to fan out grandchildren.
   */
  async branchSandbox(
    sandboxId: string,
    options: BranchOptions = {},
  ): Promise<SnapshotInfo> {
    const body: BranchOptions = {};
    if (options.tag !== undefined) body.tag = options.tag;
    // Prefer canonical `mode` when set; fall back to legacy `diff`
    // so older daemons keep working unchanged.
    if (options.mode !== undefined) {
      body.mode = options.mode;
    } else if (options.diff) {
      body.diff = true;
    }
    if (options.measure_diff) body.measure_diff = true;
    // `wait: true` is the daemon default; only send when the caller
    // opted into fire-and-forget so the body stays minimal against
    // daemons that don't recognize the field.
    if (options.wait === false) body.wait = false;
    return this.request<SnapshotInfo>(
      "POST",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}/branch`,
      body,
    );
  }

  async execCommand(
    sandboxId: string,
    args: string[],
    options: { timeoutSecs?: number } = {},
  ): Promise<ExecResult> {
    const body: ExecOptions = {
      args,
      timeout_secs: options.timeoutSecs ?? 30,
    };
    return this.request<ExecResult>(
      "POST",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}/exec`,
      body,
    );
  }

  async evalCode(sandboxId: string, code: string): Promise<EvalResult> {
    return this.request<EvalResult>(
      "POST",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}/eval`,
      { code },
    );
  }

  async pingSandbox(sandboxId: string): Promise<PingResult> {
    return this.request<PingResult>(
      "POST",
      `/v1/sandboxes/${encodeURIComponent(sandboxId)}/ping`,
    );
  }

  // --- internals ----------------------------------------------------

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = {};
    if (body !== undefined) headers["content-type"] = "application/json";
    if (this.token) headers["authorization"] = `Bearer ${this.token}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await this.fetchImpl(url, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      if (!resp.ok) {
        let parsed: unknown;
        const text = await resp.text();
        try {
          parsed = text ? JSON.parse(text) : {};
        } catch {
          parsed = text;
        }
        throw new ControllerError(resp.status, parsed, url);
      }
      // DELETE returns 204 / empty; tolerate.
      const text = await resp.text();
      if (!text) return null as T;
      return JSON.parse(text) as T;
    } finally {
      clearTimeout(timer);
    }
  }
}
